pkg/integrations/v2/app_o11y_receiver/app_o11y_receiver.go

```diff
@@ -0,0 +1,241 @@
+
+	receiverMetricsExporter := exporters.NewReceiverMetricsExporter(exporters.ReceiverMetricsExporterConfig{
+		Reg: reg,
+	})
+
+	var exp = []exporters.AppO11yReceiverExporter{
+		receiverMetricsExporter,
+	}
+
+	if len(c.ReceiverConfig.LogsInstance) > 0 {
+		logsInstance := globals.Logs.Instance(c.ReceiverConfig.LogsInstance)
+		if logsInstance == nil {
+			return nil, fmt.Errorf("logs instance \"%s\" not found", c.ReceiverConfig.LogsInstance)
+		}
+		lokiExporter := exporters.NewLogsExporter(
+			l,
+			exporters.LogsExporterConfig{
+				LogsInstance:     logsInstance,
+				Labels:           c.ReceiverConfig.LogsLabels,
+				SendEntryTimeout: c.ReceiverConfig.LogsSendTimeout,
+			},
+			sourcemapStore,
+		)
+		exp = append(exp, lokiExporter)
+	}
+
+	if len(c.ReceiverConfig.TracesInstance) > 0 {
+		tracesInstance := globals.Tracing.Instance(c.ReceiverConfig.TracesInstance)
+		if tracesInstance == nil {
+			return nil, fmt.Errorf("traces instance \"%s\" not found", c.ReceiverConfig.TracesInstance)
+		}
+
+		factory := tracesInstance.GetFactory(component.KindReceiver, pushreceiver.TypeStr)
+		if factory == nil {
+			return nil, fmt.Errorf("push receiver factory not found for traces instance \"%s\"", c.ReceiverConfig.TracesInstance)
+		}
+		tracesExporter := exporters.NewTracesExporter(factory.(*pushreceiver.Factory))
+		exp = append(exp, tracesExporter)
+	}
+
+	handler := handler.NewAppO11yHandler(c.ReceiverConfig, exp, reg)
+
+	return &appo11yIntegration{
+		logger:          l,
+		integrationName: c.Name(),
+		instanceID:      id,
+		common:          c.Common,
+		globals:         globals,
+		conf:            c.ReceiverConfig,
+		handler:         handler,
+		exporters:       exp,
+		reg:             reg,
+	}, nil
+}
+
+// Handler is the http endpoint for exposing Prometheus metrics
+func (i *appo11yIntegration) Handler(prefix string) (http.Handler, error) {
+	r := mux.NewRouter()
+	r.Handle(path.Join(prefix, "metrics"), promhttp.HandlerFor(i.reg, promhttp.HandlerOpts{}))
+	return r, nil
+}
+
+// Targets implements MetricsIntegration
+func (i *appo11yIntegration) Targets(ep integrations.Endpoint) []*targetgroup.Group {
+	integrationNameValue := model.LabelValue("integrations/" + i.integrationName)
+
+	group := &targetgroup.Group{
+		Labels: model.LabelSet{
+			model.InstanceLabel: model.LabelValue(i.instanceID),
+			model.JobLabel:      integrationNameValue,
+			"agent_hostname":    model.LabelValue(i.globals.AgentIdentifier),
+
+			// Meta labels that can be used during SD.
+			"__meta_agent_integration_name":     model.LabelValue(i.integrationName),
+			"__meta_agent_integration_instance": model.LabelValue(i.instanceID),
+		},
+		Source: fmt.Sprintf("%s/%s", i.integrationName, i.instanceID),
+	}
+
+	group.Targets = append(group.Targets, model.LabelSet{
+		model.AddressLabel:     model.LabelValue(ep.Host),
+		model.MetricsPathLabel: model.LabelValue(path.Join(ep.Prefix, "/metrics")),
+	})
+
+	return []*targetgroup.Group{group}
```

Comment:
Can we refactor this code so that it's not being duplicated across multiple integrations?

Reply:
- [ ] attempt to commonize